src/riak_core_info_service_process.erl

```diff
@@ -0,0 +1,120 @@
+%% -------------------------------------------------------------------
+%%
+%% Copyright (c) 2017 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @see riak_core_info_service
+-module(riak_core_info_service_process).
+-behaviour(gen_server).
+
+-record(state, {
+    service_provider = undefined :: riak_core_info_service:callback(),
+    response_handler = undefined :: riak_core_info_service:callback(),
+    shutdown = undefined :: riak_core_info_service:callback()
+}).
+
+-export([start_link/4,
+         init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-ifdef(TEST).
+-compile(export_all).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+%% `gen_server' implementation
+
+start_link(Registration, Shutdown, Provider, ResponseHandler) ->
+    State = #state{shutdown = Shutdown,
+                   service_provider = Provider,
+                   response_handler = ResponseHandler},
+    {ok, Pid} = gen_server:start_link(?MODULE, State, []),
+    apply_callback(Registration, [Pid]),
+    {ok, Pid}.
+
+init(State) ->
+    {ok, State}.
+
+handle_info({invoke, ProviderParams, HandlerContext}, State) ->
+    handle_invoke_message(ProviderParams, HandlerContext, State);
+
+handle_info(callback_shutdown, State) ->
```

Comment:
`handle_shutdown_message` will be called twice when handling `callback_shutdown` since it is also called in `terminate` callback. Perhaps change this clause to only return `{stop, normal, State}` and call `apply_callback(Shutdown, ...` from `terminate`?